### ISSUE TITLE
Sherloq 0.89 Issues with auto_bright_thr=0

### ISSUE THREAD

**Issue by JohnConnor01** [GuidoBartoli/sherloq#96]
I have been testing the new version of Sherloq 0.89 after the recent changes. There are some observations I have noticed since **"auto_bright_thr=0"** has been added. From what I've seen, this is making it very difficult/impossible to to separate the signal and noise. My post will be in reference to the initial post.
Initial Post: https://github.com/GuidoBartoli/sherloq/issues/86
Isolated patches were showing up on a photo, and this was being cited as areas affected by clipping. There was also a Histogram comparison being shown. As the RGB intensity values get pulled significantly to the left, it becomes harder and harder to separate the signal from the noise **for specific images**. So I wanted to adjust the settings to see if these areas still appeared when there was barely any clipping of the RGB intensity values (anything over 255 gets clipped & autobright is activated). I will show these examples below. First there are some important points to make:
1.) Adding this 1 setting, and seeing the patches go away. **Does not mean clipping was causing the patches to appear in the first place**. 
2.) These very patches were in almost **identical areas** on this same photo in **Dark Table** - **a completely separate application.** I will illustrate this below. I also did not see the original poster mention the Dark Table results. I am not sure if he was unaware, but checking other applications for corroborating results would be one of the first things I would do in a situation like this. 
3.) Autobright was not being called in Sherloq 0.87. Even though it was being cited as the reason for clipping, and subsequently the reason for patches showing up on a photo. To address this, auto_bright_thr=0 was added to Sherloq 0.89. 
This will potentially cause a wholesale shift in RGB intensity values of many photos read by Sherloq 0.89. And it will not prevent clipping with all RAW photos. 

The next point is very important 
**4.) This is also going to default to the settings that were present in the original RAW File. Because in this case, the settings in the original RAW file were the thing that caused the RGB intensity values to shift to the left significantly. If I am trying to critically analyze a RAW file, why would I also want to 100% trust the RAW file? This is another reason why it would be great to have a user option.** 

This point is also very important. 
**5.) Adding auto_bright_thr=0 will NOT prevent clipping. It will prevent clipping in the photo (atadams) shared. But it will not prevent clipping universally. And this was implemented as a way to eliminate clipping- which it will clearly not do universally as highlighted below. 
Which makes a stronger case for a user option.** 
![image](https://github.com/GuidoBartoli/sherloq/assets/173414082/6b0f597c-803d-4406-b87a-82b296f501c9)

I used the same image from the original post. And I tried 3 different settings:
auto_bright_thr=0.0015
auto_bright_thr=0.003
auto_bright_thr=0.005

As the RGB intensity values get pulled to the left - more and more noise is being introduced to the analysis. 
I will start with (auto_bright_thr=0) on Signal Separation so you can see the progression. 
![image](https://github.com/GuidoBartoli/sherloq/assets/173414082/b8d69dd8-cbde-421a-af78-62e817b59a8d)

This is what the same image looked like on Sherloq 0.87 with Signal Separation & Bit Plane.
**These same areas showed up on Dark Table** (below)- an entirely different application. 
![image](https://github.com/GuidoBartoli/sherloq/assets/173414082/822dffcf-a0a9-4067-b51d-5357a2d21651)

This is the same image compared with Sherloq 0.87 and Sherloq 0.89.
![image](https://github.com/GuidoBartoli/sherloq/assets/173414082/0cd7b00a-239a-42b7-8820-307da1f884bb)

Below. I will show auto_bright_thr=0.0015. Signal Separation, Bit Plane, Histogram. 
**As you can see even with 0.0015, the patchy areas still appear.** 
![image](https://github.com/GuidoBartoli/sherloq/assets/173414082/3a8110a8-057e-4961-ab52-e9781db76797)
Histogram on auto_bright_thr=0.0015
![image](https://github.com/GuidoBartoli/sherloq/assets/173414082/7daa7ec2-af2a-4b2e-8a10-3f8f583a836b)
**Now I will move to auto_bright_thr=0.003**
![image](https://github.com/GuidoBartoli/sherloq/assets/173414082/03f102b1-6556-4835-bfcf-023b3afd223f)
Histogram on auto_bright_thr=0.003
![image](https://github.com/GuidoBartoli/sherloq/assets/173414082/c6ccea80-9299-413f-bd4d-36e157585535)
**Lastly, this is auto_bright_thr=0.005**
![image](https://github.com/GuidoBartoli/sherloq/assets/173414082/61ef82fa-47c8-403c-8cf7-af12c6b5d87b)
Histogram on auto_bright_thr=0.005
![image](https://github.com/GuidoBartoli/sherloq/assets/173414082/545a50df-d3a6-423f-b1b2-b1587c925e06)

I mentioned earlier that adding auto_bright_thr=0 and seeing these patches go away. Does NOT mean that clipping from an autobright setting was causing the patches to appear in the first place in Sherloq 0.87. 
There is a **more plausible answer** and that is auto_bright_thr=0 is pulling some photos RGB intensity values to the left and making it very difficult/impossible to separate the signal from the noise. This is simply causing the image to be flooded with noise, making forensic analysis impossible. 
(This does not happen with ALL RAW photos. And we have already demonstrated how numerous RAW photos will still be clipped with this setting) 

**Given there is a second application, Dark Table, that shows the same areas as Sherloq 0.87**. And given how significant these changes have been to all photos analyzed with Sherloq. I would greatly appreciate if there could be a user option for this. 
(typo was corrected)

Thank you.

**Comment by BakersTutz** [comment]
The auto brightness threshold is the percentage of pixels that will be clipped _after_ brightening. The auto brightness doesn’t get “activated” after the clipping exceeds a certain amount. It’s already on and the default threshold is 1%. 

Those patches in your signal separation screenshots are the areas that are being clipped _due to the auto brightness_. This is something you DON’T want to happen. You are compromising your image data this way.

**Comment by vincent-wmtm** [comment]
This issue appears to be a duplicate of https://github.com/GuidoBartoli/sherloq/issues/93 and can be closed.

**Comment by Ray9T** [comment]
Good analysis @JohnConnor01 
I agree there is no proof of clipping resulted in the patches on the image.
auto_brightness_thr=0, in this case (due to small set of pixels out of range) is disabling auto brightness.

An erroneous and unproven proposition that the default setting 0.01 clips the image, therefore the right value is 0 has resulted in strange behavior. It may be worth noting that Rawpy does not recommend setting the value to 0 .
We are essentially going to consume the settings the potential fabricated RAW image sets ( and it's noise to hide issues).

If such a setting must be introduced, i request author to allow users to set their settings, which i'm sure image fabricators will push back.

**Comment by JohnConnor01** [comment]
> This issue appears to be a duplicate of #93 and should probably be closed. @JohnConnor01 you can add info over there if needed.

I have not seen any responses in regards to the same areas showing up on a different application? If I missed it could you please link it? 
One of the first things you would look at in a situation like this is other applications and whether or not they can corroborate the results. 

In this case. The results from Sherloq 0.87 were corroborated by Dark Table. **In the same areas.**

![image](https://github.com/GuidoBartoli/sherloq/assets/173414082/ec263ab2-a20e-4f62-bdec-48b23deaa64f)

**Comment by vincent-wmtm** [comment]
> I have not seen any responses in regards to the same areas showing up on a different application? If I missed it could you please link it?

The issue creator blocked everyone from the discussion before your darktable screenshots could be discussed.  The settings you're using in darktable are extreme and essentially creating the strange visuals you're seeing out of nothing.  You could reproduce this with just about any other image that had strong highlights.

Regardless, this is essentially the same complaint regarding the same code change and probably doesn't need its own thread.
https://github.com/GuidoBartoli/sherloq/issues/93

**Comment by JohnConnor01** [comment]
> The auto brightness threshold is the percentage of pixels that will be clipped _after_ brightening. The auto brightness doesn’t get “activated” after the clipping exceeds a certain amount. It’s already on and the default threshold is 1%.
> 
> Those patches in your signal separation screenshots are the areas that are being clipped _due to the auto brightness_. This is something you DON’T want to happen. You are compromising your image data this way.
> 
> “As the RGB intensity values get pulled to the left - more and more noise is being introduced to the analysis.” Nope. That noise is already there in the original image. When you are clipping the highlights, you are _removing_ the noise. You are not _adding_ any noise.

Is the goal of Photo Forensics to have zero clipped RGB intensity values? This is a forensics tool. Not a basic image viewer. 

Separately. You made a declarative statement in your response. “Those patches in your signal separation screenshot are the areas being clipped due to the auto brightness”

Where is your evidence of this? Because if you add the setting that makes it go away. And then the histogram shows no clipping? Is that the sum total evidence you have for this?

What would be your explanation for why these exact areas also show up on an entirely different application, Dark Table?

![image](https://github.com/GuidoBartoli/sherloq/assets/173414082/b532487d-3da3-4747-a88f-42d930ef8dde)

**Comment by JohnConnor01** [comment]
> > I have not seen any responses in regards to the same areas showing up on a different application? If I missed it could you please link it?
> 
> The issue creator blocked everyone from the discussion before your darktable screenshots could be discussed. The settings you're using in darktable are extreme and essentially creating the strange visuals you're seeing out of nothing. You could reproduce this with just about any other image that had strong highlights.
> 
> Regardless, this is essentially the same complaint regarding the same code change and probably doesn't need its own thread. #93

I would suggest corroborating results from an entirely different application is as significant as anything that was shown in the original post. In a thorough analysis. **That would be one of the first things analyzed.**

I’m sure there are other changes that can be made to the code to make those patches go away. But it also doesn’t mean the thing that was changed - causes those patches in the first place. 
More thorough investigation and research is always needed. 

**For the record. Please see the response about corroborating results from other applications.** 

![image](https://github.com/GuidoBartoli/sherloq/assets/173414082/6673c057-5cc6-4296-b959-a2d263d837c8)

**Comment by TJPofTexas** [comment]
I have confirmed what @vincent-wmtm is saying about the issue @JohnConnor01 is observing in darktable by loading the image in question and presenting the result below.

It seems to be a severe misapplication of black levels during import which is causing considerable clipping on @JohnConnor01's image.

See example below of IMG_1839 taken by Jonas De Ro, which exhibits no clipping with default CR2 raw load settings in darktable.

![image](https://github.com/GuidoBartoli/sherloq/assets/173004318/90a2972f-bc4b-4497-953e-b1c9d78408d4)

Further, if this image is then saved as jpeg and then imported in Sherloq (even in 0.87C), it does not exhibit this clipping observed by @JohnConnor01.

![image](https://github.com/GuidoBartoli/sherloq/assets/173004318/ecc34232-f754-4113-8303-44b08c794c01)

This seems to be a case of user error and incorrect settings by @JohnConnor01

**Comment by PeterWem** [comment]
> > This issue appears to be a duplicate of #93 and should probably be closed. @JohnConnor01 you can add info over there if needed.
> 
> I have not seen any responses in regards to the same areas showing up on a different application? If I missed it could you please link it? One of the first things you would look at in a situation like this is other applications and whether or not they can corroborate the results.
> 
> In this case. The results from Sherloq 0.87 were corroborated by Dark Table. **In the same areas.**
> 
> ![image](https://private-user-images.githubusercontent.com/173414082/342017784-ec263ab2-a20e-4f62-bdec-48b23deaa64f.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkxMTIwMjQsIm5iZiI6MTcxOTExMTcyNCwicGF0aCI6Ii8xNzM0MTQwODIvMzQyMDE3Nzg0LWVjMjYzYWIyLWEyMGUtNGY2Mi1iZGVjLTQ4YjIzZGVhYTY0Zi5qcGVnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYyMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MjNUMDMwMjA0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTk0MDlkYzAyYmE2MGUzYzcwMDllZmZhYzEwYTA3YTYzNDYzZjU4NTY1M2Y1YzczMGM0ODY5M2ZkOWU5ZGI5MyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.O-SPxJWztPoBoqkfCMnaRaWHIcw3p37HqVqvLYC_910)

Canon uses black level 1023, 1023, 1023 and 1023 for that image. Hit reset and read the manual. 

The darktable developers locked the two threads the user MediaTruth created. This due to harassment from MediaTruth.

https://discuss.pixls.us/t/verifying-whether-this-image-has-been-doctored/41894/

https://discuss.pixls.us/t/odds-of-subject-distance-similarities-to-a-2-decimal-point-with-a-mix-of-camera-settings/41920/2

**Comment by MediaTruth** [comment]
> > > I have not seen any responses in regards to the same areas showing up on a different application? If I missed it could you please link it?
> > 
> > 
> > The issue creator blocked everyone from the discussion before your darktable screenshots could be discussed. The settings you're using in darktable are extreme and essentially creating the strange visuals you're seeing out of nothing. You could reproduce this with just about any other image that had strong highlights.
> > Regardless, this is essentially the same complaint regarding the same code change and probably doesn't need its own thread. #93
> 
> I would suggest corroborating results from an entirely different application is as significant as anything that was shown in the original post. In a thorough analysis. **That would be one of the first things analyzed.**
> 
> I’m sure there are other changes that can be made to the code to make those patches go away. But it also doesn’t mean the thing that was changed - causes those patches in the first place. More thorough investigation and research is always needed.
> 
> **For the record. Please see the response about corroborating results from other applications.**
> 
> ![image](https://private-user-images.githubusercontent.com/173414082/342018523-6673c057-5cc6-4296-b959-a2d263d837c8.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkxMTI4MDcsIm5iZiI6MTcxOTExMjUwNywicGF0aCI6Ii8xNzM0MTQwODIvMzQyMDE4NTIzLTY2NzNjMDU3LTVjYzYtNDI5Ni1iOTU5LWEyZDI2M2Q4MzdjOC5qcGVnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYyMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MjNUMDMxNTA3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTRhNjQ4ODJkYmE4MmMxOTBlNjE2MDZlZmU2MjNmM2NkNTg2ZDA4NTAzZWFmYWY5ODJiNmE2NmQ3MmYwOTk3YSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.U7syuYF7NDOnA-gS-FS5rkaVXtEGc9jOqcIaxi2JTwA)

Interesting.
Your images show no significant changes in patch areas as tou changed the settings from 0.005 to 0015. But somehow 0 is hiding these patches with additional noise. 

Very intriguing, and a demands additional investigation.

**Comment by JohnConnor01** [comment]
Please see all 4 photos with this mountain from the file we are discussing. Given that this specific area is the area in the original photo that has the patches, I wanted to analyze this area/asset on the other 3 photos. 
If what we see in the original photo is because of clipping. I would not expect to see this same area show issues in the other 3 photos.
I am using Sherloq 0.89. And I will set the auto_bright_thru=0.005 for all 4. 
Link to file with 19 photos  https://drive.google.com/file/d/1JT0KOI1yJEtZVzdQtVBHWzyKujFDlBrb/view

**Original Photo 1837**
![image](https://github.com/GuidoBartoli/sherloq/assets/173414082/fe64f07d-e708-4bd0-9e12-c4b12fffdfe8)

1839
![image](https://github.com/GuidoBartoli/sherloq/assets/173414082/d23c3726-3e82-4a49-82f1-a94fe0ffa5de)

1840
![image](https://github.com/GuidoBartoli/sherloq/assets/173414082/45283da5-c75d-422f-9f39-39de6a71475b)

1841
![image](https://github.com/GuidoBartoli/sherloq/assets/173414082/53266eaa-01b3-4c81-876c-40b309d6b947)

**Comment by JohnConnor01** [comment]
> > > This issue appears to be a duplicate of #93 and should probably be closed. @JohnConnor01 you can add info over there if needed.
> > 
> > 
> > I have not seen any responses in regards to the same areas showing up on a different application? If I missed it could you please link it? One of the first things you would look at in a situation like this is other applications and whether or not they can corroborate the results.
> > In this case. The results from Sherloq 0.87 were corroborated by Dark Table. **In the same areas.**
> > ![image](https://private-user-images.githubusercontent.com/173414082/342017784-ec263ab2-a20e-4f62-bdec-48b23deaa64f.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkxMTIwMjQsIm5iZiI6MTcxOTExMTcyNCwicGF0aCI6Ii8xNzM0MTQwODIvMzQyMDE3Nzg0LWVjMjYzYWIyLWEyMGUtNGY2Mi1iZGVjLTQ4YjIzZGVhYTY0Zi5qcGVnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYyMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MjNUMDMwMjA0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTk0MDlkYzAyYmE2MGUzYzcwMDllZmZhYzEwYTA3YTYzNDYzZjU4NTY1M2Y1YzczMGM0ODY5M2ZkOWU5ZGI5MyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.O-SPxJWztPoBoqkfCMnaRaWHIcw3p37HqVqvLYC_910)
> 
> Canon uses black level 1023, 1023, 1023 and 1023 for that image. Hit reset and read the manual.
> 
> The darktable developers locked the two threads the user MediaTruth created. This due to harassment from MediaTruth.
> 
> https://discuss.pixls.us/t/verifying-whether-this-image-has-been-doctored/41894/
> 
> https://discuss.pixls.us/t/odds-of-subject-distance-similarities-to-a-2-decimal-point-with-a-mix-of-camera-settings/41920/2

Hello. Well I am not MediaTruth so I can't speak for other people. I am simply addressing an issue about an open source forensic tool changing a setting. 

But nothing said in your post addresses the fact the Dark Table the software showed areas in the same places Sherloq 0.87 did. So a different application corroborated the results. The same results that were rejected on Sherloq, and the reason cited was clipping. Clipping was causing the patches to show up. 

And I just analyzed all 4 photos in the same area. Can you please see my last response and reply to that.

**Comment by MediaTruth** [comment]
> > > This issue appears to be a duplicate of #93 and should probably be closed. @JohnConnor01 you can add info over there if needed.
> > 
> > 
> > I have not seen any responses in regards to the same areas showing up on a different application? If I missed it could you please link it? One of the first things you would look at in a situation like this is other applications and whether or not they can corroborate the results.
> > In this case. The results from Sherloq 0.87 were corroborated by Dark Table. **In the same areas.**
> > ![image](https://private-user-images.githubusercontent.com/173414082/342017784-ec263ab2-a20e-4f62-bdec-48b23deaa64f.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkxMTIwMjQsIm5iZiI6MTcxOTExMTcyNCwicGF0aCI6Ii8xNzM0MTQwODIvMzQyMDE3Nzg0LWVjMjYzYWIyLWEyMGUtNGY2Mi1iZGVjLTQ4YjIzZGVhYTY0Zi5qcGVnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYyMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MjNUMDMwMjA0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTk0MDlkYzAyYmE2MGUzYzcwMDllZmZhYzEwYTA3YTYzNDYzZjU4NTY1M2Y1YzczMGM0ODY5M2ZkOWU5ZGI5MyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.O-SPxJWztPoBoqkfCMnaRaWHIcw3p37HqVqvLYC_910)
> 
> Canon uses black level 1023, 1023, 1023 and 1023 for that image. Hit reset and read the manual.
> 
> The darktable developers locked the two threads the user MediaTruth created. This due to harassment from MediaTruth.
> 
> https://discuss.pixls.us/t/verifying-whether-this-image-has-been-doctored/41894/
> 
> https://discuss.pixls.us/t/odds-of-subject-distance-similarities-to-a-2-decimal-point-with-a-mix-of-camera-settings/41920/2

I 've nothing to do with that post and don't appreciate these constant personal attacks and spamming.

**Comment by JohnConnor01** [comment]
> I have confirmed what @vincent-wmtm is saying about the issue @JohnConnor01 is observing in darktable by loading the image in question and presenting the result below.
> 
> It seems to be a severe misapplication of black levels during import which is causing considerable clipping on @JohnConnor01's image.
> 
> See example below of IMG_1839 taken by Jonas De Ro, which exhibits no clipping with default CR2 raw load settings in darktable.
> 
> ![image](https://private-user-images.githubusercontent.com/173004318/342018282-90a2972f-bc4b-4497-953e-b1c9d78408d4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkxMTUwNTYsIm5iZiI6MTcxOTExNDc1NiwicGF0aCI6Ii8xNzMwMDQzMTgvMzQyMDE4MjgyLTkwYTI5NzJmLWJjNGItNDQ5Ny05NTNlLWIxYzlkNzg0MDhkNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYyM1QwMzUyMzZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02ZTRlNTUzZTU1YWUxNDMwMTgxNzFkMjRhMGIzYjM3YTZmMTNmOWYyMzg2OGQ4MGMyMjIzZjJkMjM2MzI2YjU1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.nQKE0HcXfgzOJnPzy3fJXQxaE8tgmkvQxKPSSs3TAJA)
> 
> Further, if this image is then saved as jpeg and then imported in Sherloq (even in 0.87C), it does not exhibit this clipping observed by @JohnConnor01.
> 
> ![image](https://private-user-images.githubusercontent.com/173004318/342018589-ecc34232-f754-4113-8303-44b08c794c01.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkxMTUwNTYsIm5iZiI6MTcxOTExNDc1NiwicGF0aCI6Ii8xNzMwMDQzMTgvMzQyMDE4NTg5LWVjYzM0MjMyLWY3NTQtNDExMy04MzAzLTQ0YjA4Yzc5NGMwMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYyM1QwMzUyMzZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kZGJlMDUyOGU5MGQ4M2MxOTBhMDE2M2MxNjI1YjdiYjM1NmQ1YjZhMDcxMmQ3YTBjMzUwNTViMjQzY2RhYmFlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.N4K2DcxpSK3Bdq3xAxyHlOWNhv-USb3tWy8NSpwumX8)
> 
> This seems to be a case of user error and incorrect settings by @JohnConnor01


**For the record. Please see that Dark Table is also being cited for clipping in the same areas.**

**Comment by TJPofTexas** [comment]
Are you BobbyO_ from Twitter / X?

**Comment by JohnConnor01** [comment]
**For the rest of the thread 

Can we please keep the conversation 100% focused on the substance at hand.**

Specific changes were made to an open source software tool that impacts the way it works. I’m simply trying to understand the case for:

**The clipping on 0.87 was causing the patches/areas to show up on the original photo 1837.**

**Comment by GreatRevealer** [comment]
Another used on X seems to have had the exact same issue, with the exact same photos and everyone was able to show him how the problem was not with the images, but with how Sherloq loaded CR2 RAW images.

These posts will probably help you solve your problem also.

https://x.com/BobbyO_/status/1803462039634952408

This user WSAdvisor_ was ALSO having the exact same issues with the exact same images, so this may also help you...

https://x.com/WSAdvisor_/status/1793731789649019342

**Comment by JohnConnor01** [comment]
> Another used on X seems to have had the exact same issue, with the exact same photos and everyone was able to show him how the problem was not with the images, but with how Sherloq loaded CR2 RAW images.
> 
> These posts will probably help you solve your problem also.
> 
> https://x.com/BobbyO_/status/1803462039634952408
> 
> This user WSAdvisor_ was ALSO having the exact same issues with the exact same images, so this may also help you...
> 
> https://x.com/WSAdvisor_/status/1793731789649019342

I do appreciate the information but unfortunately that doesn’t address this issue. This is about auto_bright_thru=0 and the new settings on Sherloq. I’ll copy/paste the earlier post. But thank you for sharing. 
For the rest of the thread. Can we please keep the conversation 100% focused on the substance at hand.
Specific changes were made to an open source software tool that impacts the way it works. I’m simply trying to understand the case for:

**The clipping on 0.87 was causing the patches/areas to show up on the original photo 1837.** 

I am showing information and analysis that anyone can replicate. And asking, IF it was the clipping that was causing the issue in 1837. Would we expect to be seeing any issues with the same area in the other photos? 
Given those are different photos.

**Comment by JohnConnor01** [comment]
I appreciate any thoughtful analysis of the conversation at hand. 

I am noticing alot of comments that aren’t addressing the issues I am raising. For that I may have to moderate the thread. 

Hopefully we can stay focused on auto_bright_thru=0 and whether or not it makes sense to have a user option given how significant the changes are. And how little supporting evidence was originally presented. 

I would appreciate anyone’s thoughts on the 4 different photos on 0.89 with auto_bright_thru=0.005

**Comment by GreatRevealer** [comment]
The clipping issue is caused by auto bright.  

If you manually turn off auto_bright, the clipping issue will go away completely.

As you increase the threshold clipping ratio, more and more pixels are allowed to be clipped, as observed in your four test photos.

If you import the raw photo through darktable without manipulating black levels and export a jpeg, then import that to Sherloq, you will also see there is no clipping.

It appears the clipping is an error you are introducing in darktable by dramatically changing the black levels.

![image](https://github.com/GuidoBartoli/sherloq/assets/173382606/fb19627c-c52d-4f86-afeb-6adcc3ce46d7)

What made you change these settings in this way?  This is a very unusual change to make.

Also, are these images part of some photo competition?  I see everyone working on the same photos like maybe it's a school project.

Are you all working together or against each other?  Hopefully you're not breaking your teachers rules by cheating to get answers here...

**Comment by PeterWem** [comment]
> > > > This issue appears to be a duplicate of #93 and should probably be closed. @JohnConnor01 you can add info over there if needed.
> > > 
> > > 
> > > I have not seen any responses in regards to the same areas showing up on a different application? If I missed it could you please link it? One of the first things you would look at in a situation like this is other applications and whether or not they can corroborate the results.
> > > In this case. The results from Sherloq 0.87 were corroborated by Dark Table. **In the same areas.**
> > > ![image](https://private-user-images.githubusercontent.com/173414082/342017784-ec263ab2-a20e-4f62-bdec-48b23deaa64f.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkxMTIwMjQsIm5iZiI6MTcxOTExMTcyNCwicGF0aCI6Ii8xNzM0MTQwODIvMzQyMDE3Nzg0LWVjMjYzYWIyLWEyMGUtNGY2Mi1iZGVjLTQ4YjIzZGVhYTY0Zi5qcGVnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYyMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MjNUMDMwMjA0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTk0MDlkYzAyYmE2MGUzYzcwMDllZmZhYzEwYTA3YTYzNDYzZjU4NTY1M2Y1YzczMGM0ODY5M2ZkOWU5ZGI5MyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.O-SPxJWztPoBoqkfCMnaRaWHIcw3p37HqVqvLYC_910)
> > 
> > 
> > Canon uses black level 1023, 1023, 1023 and 1023 for that image. Hit reset and read the manual.
> > The darktable developers locked the two threads the user MediaTruth created. This due to harassment from MediaTruth.
> > https://discuss.pixls.us/t/verifying-whether-this-image-has-been-doctored/41894/
> > https://discuss.pixls.us/t/odds-of-subject-distance-similarities-to-a-2-decimal-point-with-a-mix-of-camera-settings/41920/2
> 
> I 've nothing to do with that post and don't appreciate these constant personal attacks and spamming.

So why did you right now upload the picture of the sunset, original posted by the harasser?
![3_Orbs-Cr2-TIFF-DNG](https://github.com/GuidoBartoli/sherloq/assets/34200649/4b6bc63b-af9a-441b-997f-32937bcf1805)

**Comment by JohnConnor01** [comment]
I am getting more and more comments that don’t address the issues. 

I analyzed all 4 photos in the same area on Sherloq 0.89. With auto_bright_thru=0.005. 

**Please address that if you are doing to comment.**

**Comment by JohnConnor01** [comment]
> > > > > This issue appears to be a duplicate of #93 and should probably be closed. @JohnConnor01 you can add info over there if needed.
> > > > 
> > > > 
> > > > I have not seen any responses in regards to the same areas showing up on a different application? If I missed it could you please link it? One of the first things you would look at in a situation like this is other applications and whether or not they can corroborate the results.
> > > > In this case. The results from Sherloq 0.87 were corroborated by Dark Table. **In the same areas.**
> > > > ![image](https://private-user-images.githubusercontent.com/173414082/342017784-ec263ab2-a20e-4f62-bdec-48b23deaa64f.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkxMTIwMjQsIm5iZiI6MTcxOTExMTcyNCwicGF0aCI6Ii8xNzM0MTQwODIvMzQyMDE3Nzg0LWVjMjYzYWIyLWEyMGUtNGY2Mi1iZGVjLTQ4YjIzZGVhYTY0Zi5qcGVnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYyMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MjNUMDMwMjA0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTk0MDlkYzAyYmE2MGUzYzcwMDllZmZhYzEwYTA3YTYzNDYzZjU4NTY1M2Y1YzczMGM0ODY5M2ZkOWU5ZGI5MyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.O-SPxJWztPoBoqkfCMnaRaWHIcw3p37HqVqvLYC_910)
> > > 
> > > 
> > > Canon uses black level 1023, 1023, 1023 and 1023 for that image. Hit reset and read the manual.
> > > The darktable developers locked the two threads the user MediaTruth created. This due to harassment from MediaTruth.
> > > https://discuss.pixls.us/t/verifying-whether-this-image-has-been-doctored/41894/
> > > https://discuss.pixls.us/t/odds-of-subject-distance-similarities-to-a-2-decimal-point-with-a-mix-of-camera-settings/41920/2
> > 
> > 
> > I 've nothing to do with that post and don't appreciate these constant personal attacks and spamming.
> 
> So why did you right now upload the picture of the sunset, original posted by the harasser? ![3_Orbs-Cr2-TIFF-DNG](https://private-user-images.githubusercontent.com/34200649/342023970-4b6bc63b-af9a-441b-997f-32937bcf1805.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkxMTg0NDAsIm5iZiI6MTcxOTExODE0MCwicGF0aCI6Ii8zNDIwMDY0OS8zNDIwMjM5NzAtNGI2YmM2M2ItYWY5YS00NDFiLTk5N2YtMzI5MzdiY2YxODA1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjIzVDA0NDkwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc0OWUyYTg3ZDlkMWVhNjljMTg2YTJiNmUyMGQxZjUzZmJjODY0OTQzMTgyYzQ5ZTMwMGYwNjVkYzAxZDAzMDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Y46hL-Yr02bIoxqCzmbZ6qHyjDknIIZC7epKTfvqWuw)



I have no idea what that is but I blocked this user for spamming. 

Please address the 4 photo analysis on Sherloq 0.89. Or something about the issue being discussed. Auto_bright_thru=0

**Comment by BakersTutz** [comment]
There is absolute no reason why anyone would allow the highlights of an image to clip on purpose, if the intent is a forensic analysis and to determine potential photo manipulation. 

It is recommended to either turn off auto bright completely, or keep it turned on with the threshold set to 0% clipping. You need to retain ALL of the image data if you want a complete and accurate analysis.

**Comment by SacredCowBBQ** [comment]
Alex Tutubalin, a representative for LiBRaw LLC, the maker of the RawPy library which wrote raw.postprocess(), the function in question, recommends using 'auto_bright_thr' set to 0 to avoid clipping over over-exposed images.

This is exactly the update that has been applied to Sherloq in 0.89d.  This is a good change, and one worth keeping.

See their post here: 
https://www.libraw.org/node/2267

_______________________________________________________

Now that that is out of the way!

@JohnConnor01 / BobbyO_, everyone can see right through you and knows what you are doing!

https://x.com/BobbyO_/status/1803462039634952408
![image](https://github.com/GuidoBartoli/sherloq/assets/173579287/28d1cd06-dd9e-4080-add1-cb62f70da5e6)

These are the exact same images you posted to Twitter!

You keep trying to side step this.

What you are doing is trying to harrass and defame Jonas De Ro by fraudulently and deceptively producing evidence that his images are fake by using an EDUCATIONAL TOOL that is unofficial!

You are blocking EVERYONE and ignoring them when the provide the answer you can wait on Ray9T, MediaTruth, or NaturalML to come agree with you.

**Comment by JohnConnor01** [comment]
I’ll share the 4 photo analysis again. It is important to emphasize these are 4 different photos taken at different times, angles, different lighting, etc etc.

**And the explanation is clipping.**

![image](https://github.com/GuidoBartoli/sherloq/assets/173414082/577bfb87-f7a9-4405-b8bc-cbfdca020445)

![image](https://github.com/GuidoBartoli/sherloq/assets/173414082/ebe9b630-30eb-46a6-9338-9fda630d6484)

![image](https://github.com/GuidoBartoli/sherloq/assets/173414082/606bc1b4-69cf-4337-b55e-eaa19e598321)

![image](https://github.com/GuidoBartoli/sherloq/assets/173414082/9c28a981-bc34-4e40-809d-5a66101a251e)


Please respond to this specifically or I have to moderate **There is too much spam.**

**Comment by JohnConnor01** [comment]
> There is absolutely no reason why anyone would allow the highlights of an image to clip on purpose, if the intent is to conduct a forensic analysis and to determine potential photo manipulation.
> 
> It is recommended to either turn off auto bright completely, or keep it turned on with the threshold set to 0% clipping. You need to retain ALL of the image data if you want a complete and accurate analysis.



**Please see repeatedly not addressing the 4 photo analysis.**

THERE IS TOO MUCH SPAM

**Comment by JohnConnor01** [comment]
BLOCKING ALL SPAM

This is a serious issue and an original issue was opened on this. The image being presented by the original poster. Is photo 1837. So I looked at the other photos in that file because in this case it would be very relevant **given the explanation is clipping.** 
**Everything about that would be relevant.**

Please only about

**Comment by ICPosse** [comment]
MORE PAIN IS COMING 
![IMG_0862](https://github.com/GuidoBartoli/sherloq/assets/173592009/a040d883-e156-45d3-a800-a8f12becf7a2)

**Comment by JohnConnor01** [comment]
I do want to highlight some important details. 

1.) The original post that cited the issue with Sherloq 0.87 referenced the photo 1837. 
2.) The original poster attributed the parches around the Mountain to clipping.
3.) He then advocated for adding auto_bright_thru=0 because after he used that command, the patches on the Mountain go away. 

There are other explanations for why the patches go away after adding that setting. And seeing them go away after adding it is not proof that clipping was causing them to begin with. 

The original post/author introduced photo 1837 and used it as an example to suggest Sherloq 0.87 was clipping the photos and that specifically was causing patches to show up around the mountain. 

So given there are 4 photos with this same area in it. Analyzing the same area in other photos would be a great way to test this thesis.

Everything in my post is relevant to the original post made.

**Comment by JohnConnor01** [comment]
Please keep all future comments about the 4 photo analysis. The explanation that was given in the original photo - clipping. Clipping in general on a photo, where that happens, and how that evolves
photo to photo based on numerous variables. And the new setting that significantly changes the way the application operates auto_bright_thru=0. That was originally introduced by the first post/author in part because photo 1837 was showing patches on this area. 

![image](https://github.com/GuidoBartoli/sherloq/assets/173414082/f2cbce0e-fbac-4285-9405-58eebb9681c7)
![image](https://github.com/GuidoBartoli/sherloq/assets/173414082/9ea5100b-e713-41b0-a3fe-a8468d80a32c)

**Comment by atadams** [comment]
I wrote a script to output the pixels clipped by rawpy’s auto_bright. Using the default setting (auto_bright_thr=0.01) on IMG_1837 from my issue 86 post, the white pixels in the image below are clipped from the red channel.

![IMG_1837_clipped_pixels_red_010](https://github.com/GuidoBartoli/sherloq/assets/392510/d39aa04d-3107-43b4-a94d-aff8237fd5b7)

This is the clipped pixels compared to the Bit Plane Value tool.

![IMG_1937_clippedvsplane](https://github.com/GuidoBartoli/sherloq/assets/392510/6f743bb5-1595-459c-9303-4d4f236b7f72)

And the Error Level Analysis tool

![IMG_1937_clippedvsela](https://github.com/GuidoBartoli/sherloq/assets/392510/63ebecab-29a0-46a2-bb0c-c4ae2b415c0e)

Both tools are just showing the clipped pixels.

If we animate the auto_bright_thr setting from 0 to 0.01 in 0.001 steps, the relationship between the clipped pixels and these tools is even more visible.

**Bit Plane Value**
![IMG_1837_plane-sm](https://github.com/GuidoBartoli/sherloq/assets/392510/70bdfffc-8627-4c74-9356-6ca5a7eb2021)

**Error Level Analysis**
![IMG_1837_ela-sm](https://github.com/GuidoBartoli/sherloq/assets/392510/7c3657e5-f171-4e65-aaf5-1fdb9b504422)

You will see similar results in Sherloq's other tools, e.g., Signal Separation.
 
I can't imagine why anyone would want these flawed results unless they were intentionally trying to make images look manipulated. Another risk of having auto_bright_thr as an option is an inexperienced user might turn it on without knowing the ramifications and draw false conclusions about any image they view.

I see no reason for a forensic app to allow clipping an image in this way. My recommendation is to _not_ make `auto_bright_thr=0` an option. It will only give a tool to the malicious user to misuse Sherloq or confuse the naive user.

### Personal Note
It's obvious that users who are not acting in good faith have been very active on this repo recently. @GuidoBartoli has done a remarkable job with this app. There isn't another open-source tool like it. He does not deserve to have to deal with this type of activity.

_**I appeal to users not acting in good faith to leave this project alone.**_

**Comment by JohnConnor01** [comment]
In reference to the @atadams post. 

I created this account to address this issue specifically. My account being relatively new **should have no impact** on the information and analysis I am presenting. It should be 100% about the substance being discussed. 

@atadams brought up this issue originally. He cited clipping in the photo 1837. And he suggested auto_bright_thru=0 would be a good default setting for the entire application. 

To reiterate a couple earlier points about the original post.
1.) Showing a Histogram channel comparison before and after auto_bright_thru=0, and showing how the patches go away after the setting is applied. **Is not proof that clipping was causing the patches to appear in the first place.**
I am simply trying to do a further analysis on whether or not the explanation given (clipping) makes sense based on much of the information I have reviewed in this thread.
2.) There are other explanations for why the patches go away after auto_bright_thru=0 as I have highlighted in this post. 

I am clearly trying to explore **all options. Which is what someone would do who was acting in good faith.** 
And test this thesis of "clipping is causing this to happen". Which is why I referenced results from Dark Table because this was being attributed to a setting in Sherloq (moreso Rawpy) - and corroborated results from a different application would be one of the first things someone would analyze. I also analyzed the same area on all 4 photos that have that area in it. 

**I would appreciate if all future posts reference the specific information shown in this thread.**

**Comment by JohnConnor01** [comment]
Thank you for posting this @MediaTruth. 

If the purpose of the setting auto_bright_thru=0 is to eliminate the clipping of pixels with Sherloq 0.89. **This setting will clearly not do that** as highlighted below. It may do this for the images @atadams presented. 

But this setting will NOT do that universally. **Which makes the case for a user option that much stronger.**

![image](https://github.com/GuidoBartoli/sherloq/assets/173414082/86f1223c-e054-488d-a01e-744b998c7375)

**Comment by atadams** [comment]
> You are yet to prove the clipped areas are corelated to the patches. It's clear how you are avoiding this ask

I think @GuidoBartoli is smart enough to see exactly what is going on.

**Comment by JohnConnor01** [comment]
> > You are yet to prove the clipped areas are corelated to the patches. It's clear how you are avoiding this ask
> 
> I think @GuidoBartoli is smart enough to see exactly what is going on.

I am simply exploring all possibilities given how significant the changes have been. 

Given 1837 was the photo you presented as an example of clipping.

**Comment by JohnConnor01** [comment]
> > You are yet to prove the clipped areas are corelated to the patches. It's clear how you are avoiding this ask
> 
> I think @GuidoBartoli is smart enough to see exactly what is going on.


I am simply trying to test the thesis: Clipping is causing the patches to appear on the photo in 1837. And I am exploring all possibilities which is what someone would do who is acting in good faith. 

There are other explanations for why the patches go away after auto_bright_thru=0, and I have highlighted those in this thread. None of the other explanations have been addressed/explored publicly when discussing this issue from what I can see. 
1.) Adding auto_bright_thru=0 and seeing the patches go away. **Is not proof that clipping caused the patches to show up in the first place.**
2.) This setting WILL NOT prevent clipping as highlighted in the earlier post. It will prevent clipping in the specific photo @atadams shared

**Comment by JohnConnor01** [comment]
Please see the explanation given by @atadams for why there should be NO user option. 

But adding auto_bright_thru=0 is NOT a universal solution for clipping as highlighted in this thread.

So this setting is being advocated for as a way to prevent clipping, but it will only do that for specific photos - like the ones referenced by @atadams. 

This is why a user option is important. Other important reasons have also been cited in this thread. 

![image](https://github.com/GuidoBartoli/sherloq/assets/173414082/de41ee06-3a79-4666-a228-7022250bbe40)

But as shown this will NOT universally prevent clipping. 
![image](https://github.com/GuidoBartoli/sherloq/assets/173414082/4db58874-a000-4666-8e84-16608d6ed8a9)

**Comment by JohnConnor01** [comment]
To further illustrate how auto_bright_thru=0 is not a big picture solution for clipping. Here are two additional examples of RAW photos being clipped on the new settings. Auto_bright_thru=0 is activated on both photos analysis. 

Photo 1: 
![image](https://github.com/GuidoBartoli/sherloq/assets/173414082/47704d44-0cbf-431a-bd18-1958672c72a8)
Photo 2: 
![image](https://github.com/GuidoBartoli/sherloq/assets/173414082/d0b431fd-dc02-478e-8f77-6ac40f0d8335)

Analysis of Photo 1 on Sherloq 0.89 with auto_bright_thru=0
![image](https://github.com/GuidoBartoli/sherloq/assets/173414082/abdb097f-f2da-4ef2-a04e-2b92c0dc7d2d)

Analysis of Photo 2 on Sherloq 0.89 with auto_bright_thru=0
![image](https://github.com/GuidoBartoli/sherloq/assets/173414082/f7bf1375-a602-4700-aa81-079ba739a3a0)

**Comment by JohnConnor01** [comment]
Another example of a RAW photo being clipped on Sherloq 0.89 with auto_bright_thru=0 

All we are asking for is a user option given how heavily auto_bright_thru=0 impacts the analysis process. If the reason this was being implemented was to prevent RAW photos from being clipped - we have illustrated how this command **will clearly not prevent clipping**. It will prevent clipping for the original image(s) @atadams presented. **But it is not a solution that will prevent clipping for other RAW photos.** 

@atadams would appreciate your feedback on these issues.

Link to image: 
![image](https://github.com/GuidoBartoli/sherloq/assets/173414082/4c11adca-62f1-4adb-aa36-c2fa75ab6fef)

Clipping on RAW Photo
![image](https://github.com/GuidoBartoli/sherloq/assets/173414082/50df6bf1-69a7-4a64-92e6-761c2caf2795)

**Comment by atadams** [comment]
> Two more examples of a RAW photos being clipped on Sherloq 0.89 with auto_bright_thru=0
> 
> All we are asking for is a user option given how heavily auto_bright_thru=0 impacts the analysis process. If the reason this was being implemented was to prevent RAW photos from being clipped - we have illustrated how this command **will clearly not prevent clipping**. It will prevent clipping for the original image(s) @atadams presented. **But it is not a solution that will prevent clipping for other RAW photos.**

The camera’s sensor clipped your example images. That is the data the camera captured. Nothing is lost. That situation completely differs from rawpy clipping an image and _throwing out existing data_. 

I can’t see any legitimate reason why a user doing an honest image forensics evaluation would want to lose data when opening a file.

**Comment by Ray9T** [comment]
> > Two more examples of a RAW photos being clipped on Sherloq 0.89 with auto_bright_thru=0
> > All we are asking for is a user option given how heavily auto_bright_thru=0 impacts the analysis process. If the reason this was being implemented was to prevent RAW photos from being clipped - we have illustrated how this command **will clearly not prevent clipping**. It will prevent clipping for the original image(s) @atadams presented. **But it is not a solution that will prevent clipping for other RAW photos.**
> 
> The camera’s sensor clipped your example images. That is the data the camera captured. Nothing is lost. That situation completely differs from rawpy clipping an image and _throwing out existing data_.
> 
> I can’t see any legitimate reason why a user doing an honest image forensics evaluation would want to lose data when opening a file.

Then focus on getting the full data, and not brightness hacks Tony.

**Comment by atadams** [comment]
> Then focus on getting the full data, and not brightness hacks Tony.

That's what `auto_bright_thr=0` does. It makes sure you get the full data and none is thrown out.

**Comment by JohnConnor01** [comment]
> > Two more examples of a RAW photos being clipped on Sherloq 0.89 with auto_bright_thru=0
> > All we are asking for is a user option given how heavily auto_bright_thru=0 impacts the analysis process. If the reason this was being implemented was to prevent RAW photos from being clipped - we have illustrated how this command **will clearly not prevent clipping**. It will prevent clipping for the original image(s) @atadams presented. **But it is not a solution that will prevent clipping for other RAW photos.**
> 
> The camera’s sensor clipped your example images. That is the data the camera captured. Nothing is lost. That situation completely differs from rawpy clipping an image and _throwing out existing data_.
> 
> I can’t see any legitimate reason why a user doing an honest image forensics evaluation would want to lose data when opening a file.


Legitimate Reasons why a user doing an honest image forensics evaluation would NOT want auto_bright_thru=0 as a default setting with no option for users. 

1.) Showing histogram/signal separation results before/after auto_bright_thru=0. And highlighting how the patches go away after auto_bright_thru=0, is NOT proof that clipping caused the patches to appear. There are other explanations which have been highlighted in this thread.
2.) This is also going to default to the settings that were present in the original RAW File. Because in this case, the settings in the original RAW file were the thing that caused the RGB intensity values to shift to the left significantly. If I am trying to critically analyze a RAW file, why would I also want to 100% trust the RAW file? This is another reason why it would be great to have a user option.
3.) There are already MANY other RAW photos being clipped with the new version 0.89 after auto_bright_thru=0 has been added. I acknowledge atadams answer above, but he is advocating for NO user option because of clipping. When tons of RAW photos will still be clipped on 0.89 after the recent changes.
4.) These results were corroborated by at least one other (completely different application) Dark Table
5.) The lack of supporting evidence for the initial issue opened by @atadams. Auto_bright_thru=0 is now a default setting on Sherloq 0.89. So ALL photos analysis will be impacted by these changes. When there was very little supporting evidence for the changes.
**6.) Apparently now there is a 3rd application corroborating the results: Raw Therapee. This is significant in that we have TWO different applications corroborating the results.**

**Comment by atadams** [comment]
> Raw Therapee also shows the same spots on your image

The fact that you can manipulate the image in another app to highlight the pixels rawpy was clipping is not surprising. 

This argument is nonsense.

**Comment by Ray9T** [comment]
> > Then focus on getting the full data, and not brightness hacks Tony.
> 
> That's what `auto_bright_thr=0` does. It makes sure you get the full data and none is thrown out.

It only does that for your image, it's not the intended purpose of the setting.  No one should disable critical auto brightness scaler to get full data, that is wrong. 
For example, 
`raw.raw_image_visible ` gives you accesses to the full visible RAW image data. Brightness hack only works for your image.

**Comment by JohnConnor01** [comment]
> > Raw Therapee also shows the same spots on your image
> 
> The fact that you can manipulate the image in another app to highlight the pixels rawpy was clipping is not surprising.
> 
> This argument is nonsense.

If you are looking at results of a photo on a forensic application. And you believe those results are in error. 

Why wouldn't you check other applications to see if they also corroborated the results? This seems like one of the **first things** you would do?

**Comment by Ray9T** [comment]
> > > Raw Therapee also shows the same spots on your image
> > 
> > 
> > The fact that you can manipulate the image in another app to highlight the pixels rawpy was clipping is not surprising.
> > This argument is nonsense.
> 
> If you are looking at results of a photo on a forensic application. And you believe those results are in error.
> 
> Why wouldn't you check other applications to see if they also corroborated the results? This seems like one of the **first things** you would do?

It's surprising that Tony hasn't established a proof that clipping resulted in patches.  But pushed a very specific setting to **universally disable brightness** to accommodate his image patches from showing up.
Nothing sounds right about this. 
If the intent is to get full data, that should be the ask, and not the setting and it's specific values.

**Comment by JohnConnor01** [comment]
**For the record please see that TWO other applications have corroborated the results.** 

This has been dismissed as nonsense by the person who opened this issue.

**Comment by atadams** [comment]
> It only does that for your image, it's not the intended purpose of the setting. No one should disable critical auto brightness scaler to get full data, that is wrong.

Auto brightness isn't disabled. `auto_bright_thr=0` only assures no clipping (I.e., data loss) takes place when the image is auto brightness is applied.

This issue has over 40 comments. Don't you think it's been discussed enough?

**Comment by Ray9T** [comment]
> > It only does that for your image, it's not the intended purpose of the setting. No one should disable critical auto brightness scaler to get full data, that is wrong.
> 
> Auto brightness isn't disabled. `auto_bright_thr=0` only assures no clipping (I.e., data loss) takes place when auto brightness is applied.
> 
> This issue has over 40 comments. Don't you think it's been discussed enough?

> Auto brightness isn't disabled

That's an erroneous statement, are you saying auto brightness is still working? `Auto_bright_thr=0 `documentation says at this value, even if one pixel is outside the 255 range/clipped then DISABLE Auto brightness.

Now it's established that you misinterpreted the settings and applied the setting universally!

**Comment by atadams** [comment]
> **For the record please see that TWO other applications have corroborated the results.** 
> 
> 
> 
> This has been dismissed as nonsense by the person who opened this issue. 

That is only corroboration that other apps can highlight the brightest 1% of pixels, not that, from a forensic standpoint, those pixels have any significance. That is why it is nonsense.

**Comment by Ray9T** [comment]
> > **For the record please see that TWO other applications have corroborated the results.**
> > This has been dismissed as nonsense by the person who opened this issue.
> 
> That is only corroboration that other apps can highlight the brightest 1% of pixels, not that, from a forensic standpoint, those pixels have any significance. That is why it is nonsense.

Ah now you are indicating that there is a correlation between clipped pixels and patches! Which was asked many times and never provided.  I also showed you via Raw therapee how the clipped pixels are not part of the patches. 

![Screenshot 2024-06-23 081706](https://github.com/GuidoBartoli/sherloq/assets/165091653/d18bea5f-058f-4570-a2f8-675868f807f2)


Patches going away can be for many reasons,

**Comment by GuidoBartoli** [comment]
Wow, you guys wrote more than 60 comments in two days, it's really a lot and I'm just noticing it now when I get back from the weekend... My program had never received this much concentrated attention, and I don't know if that is good or bad 😆 

There is too much to read and understand to quickly respond to you knowledgeably, so loosen up for a moment and give me a couple of days to analyze it all, do some testing, and draw some conclusions, okay?

However, I thank you all anyway for taking the time to use and test Sherloq, I'm really pleased about this, it's in the spirit of open-source development!

Some initial notes based only on my perception:

@JohnConnor01, @SacredCowBBQ, @BakersTutz, @Ray9T, @vincent-wmtm, @TJPofTexas, @GreatRevealer, you all created your accounts just for commenting here and none of you has ever committed a single line of code in Github, I hope your findings are meaningful and useful, not a waste of time for the community.

@atadams, thanks for replicating results, your politeness and objectiveness is appreciated.

**Comment by JohnConnor01** [comment]
Thank you for your response @GuidoBartoli 

We are happy to create a more streamlined version of the points made. We are simply looking for a user option. I understand going through threads like this takes time and I greatly appreciate you getting back with us.

**Comment by GuidoBartoli** [comment]
Okay, here I am after analyzing this very long thread that basically just revolves around what options to choose when importing RAW files via the `rawpy` library.

Here are my conclusions on the matter:
- I have done some personal research on the people involved in this discussion and I think I have a pretty accurate idea of ​​the reality of the facts, but this in no way influences the technical decisions regarding Sherloq, which I would like to remain a valid open source tool useful for carry out research in the field of forensic analysis of digital images.
- Regarding standard raster images (JPG, PNG, TIF, ...), the program simply reads the RGB values present so all subsequent analysis is consistent with the original content. RAW images, on the other hand, contain only sensor light readings that require post-processing to obtain the "final" RGB values, which depend on many steps (demosaicing, denoising, gamma correction, white balance, black levels, and so on...).
- A forensic analysis on RAW images is in itself very rare and peculiar, because potential alterations are in almost all cases carried out by programs working in RGB, so I had initially thought of completely removing the possibility of importing RAWs, but then I felt that it could still be useful in some cases, so I decided to keep it.
- In this scenario in which post-processing of the RAW data is still necessary to obtain the RGB data on which to perform subsequent analysis, I believe that the best compromise is to reduce it to a minimum to avoid the results being even more biased.
- A premise must be made: if a photo was taken overexposed (or underexposed), balancing the exposure starting from the RAW is a common practice of photographers (myself included) to recover the areas where clipping occurred, which is not possible with 8-bit raster images due to the much smaller dynamic range.
- That said, assessing the presence of potential digital manipulation based on the RAW preprocessor's ability to recover these artifacts **makes totally no sense** from a forensic researcher's point of view, so I kindly ask you to avoid posting additional images comparing the effect of the different options on the clipped areas.
- I looked at the [rawpy default options](https://letmaik.github.io/rawpy/api/rawpy.Params.html) and saw that to minimize post-processing operations, it is necessary to disable auto brightness adjustment, auto white balance, and pixel value scaling, so please take a look at commit 5c8684a1072f6d2ea2bebb816eab914a3fd6707e where you can find these updated behavior.
